delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuMergeIntoCommand.scala

```diff
@@ -0,0 +1,1171 @@
+        createMetric(sc, "time taken to execute the entire operation"),
+    "scanTimeMs" ->
+        createMetric(sc, "time taken to scan the files for matches"),
+    "rewriteTimeMs" ->
+        createMetric(sc, "time taken to rewrite the matched files"))
+
+  override def run(spark: SparkSession): Seq[Row] = {
+    recordDeltaOperation(targetDeltaLog, "delta.dml.merge") {
+      val startTime = System.nanoTime()
+      gpuDeltaLog.withNewTransaction { deltaTxn =>
+        if (target.schema.size != deltaTxn.metadata.schema.size) {
+          throw DeltaErrors.schemaChangedSinceAnalysis(
+            atAnalysis = target.schema, latestSchema = deltaTxn.metadata.schema)
+        }
+
+        if (canMergeSchema) {
+          updateMetadata(
+            spark, deltaTxn, migratedSchema.getOrElse(target.schema),
+            deltaTxn.metadata.partitionColumns, deltaTxn.metadata.configuration,
+            isOverwriteMode = false, rearrangeOnly = false)
+        }
+
+        val deltaActions = {
+          if (isSingleInsertOnly && spark.conf.get(DeltaSQLConf.MERGE_INSERT_ONLY_ENABLED)) {
+            writeInsertsOnlyWhenNoMatchedClauses(spark, deltaTxn)
+          } else {
+            val filesToRewrite = findTouchedFiles(spark, deltaTxn)
+            val newWrittenFiles = withStatusCode("DELTA", "Writing merged data") {
+              writeAllChanges(spark, deltaTxn, filesToRewrite)
+            }
+            filesToRewrite.map(_.remove) ++ newWrittenFiles
+          }
+        }
+
+        // Metrics should be recorded before commit (where they are written to delta logs).
+        metrics("executionTimeMs").set((System.nanoTime() - startTime) / 1000 / 1000)
+        deltaTxn.registerSQLMetrics(spark, metrics)
+
+        // This is a best-effort sanity check.
+        if (metrics("numSourceRowsInSecondScan").value >= 0 &&
+            metrics("numSourceRows").value != metrics("numSourceRowsInSecondScan").value) {
+          log.warn(s"Merge source has ${metrics("numSourceRows").value} rows in initial scan but " +
+              s"${metrics("numSourceRowsInSecondScan").value} rows in second scan")
+          if (conf.getConf(DeltaSQLConf.MERGE_FAIL_IF_SOURCE_CHANGED)) {
+            throw DeltaErrors.sourceNotDeterministicInMergeException(spark)
+          }
+        }
+
+        deltaTxn.commit(
+          deltaActions,
+          DeltaOperations.Merge(
+            Option(condition.sql),
+            matchedClauses.map(DeltaOperations.MergePredicate(_)),
+            notMatchedClauses.map(DeltaOperations.MergePredicate(_)),
+            // TODO do we need to pass something in here?
+            notMatchedBySourcePredicates = Seq.empty[MergePredicate]
+        ))
+
+        // Record metrics
+        val stats = GpuMergeStats.fromMergeSQLMetrics(
+          metrics, condition, matchedClauses, notMatchedClauses,
+          deltaTxn.metadata.partitionColumns.nonEmpty)
+        recordDeltaEvent(targetDeltaLog, "delta.dml.merge.stats", data = stats)
```

Comment:
There needs to be a shim here to add the notMatchedBySourceClauses details, related to the TODO above.